### PR TITLE
D hotfix/bill

### PR DESCRIPTION
**Pull request by Lucaskabela**
Thanks for the work on this Bill - I will rebase this on the changes to Jay's branch and push the fixes (just requires passing length to an RNNEncoder so it can drop padding from its internal embedding!)

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -237,11 +237,11 @@ def supervised_step(model, batch, args, use_dis_loss=False):
     mel_aug = specaugment(mel, mel_len)
     if use_dis_loss:
         pre_pred, post_pred, stop_pred, stop_lens, t_hid = model.tts(text, text_len, mel, mel_len, ret_enc_hid=use_dis_loss)
-        t_d_loss = discriminator_hidden_to_loss(model, t_hid, 'speech', freeze_discriminator=True)
+        t_d_loss = discriminator_hidden_to_loss(model, t_hid, text_len, 'speech', args.model_type, freeze_discriminator=True)
```
Comment (Lucaskabela) [inline src/train.py]: Please check that these make sense @billyang98 - essentially these should be the ground truth lengths of the encoded state so that we know which encoder outputs correspond to padding.

File: src/train.py [hunk 2]
```diff
@@ -819,7 +814,8 @@ def initialize_model(args):
         speech_m = SpeechTransformer(args)
 
     if args.use_discriminator:
-        discriminator = Discriminator(args.hidden)
+        discriminator_in_dim = args.hidden * 2 if args.model_type == 'rnn' else args.hidden
+        discriminator = LSTMDiscriminator(discriminator_in_dim, args.disc_hid, bidirectional=args.disc_bidirectional, num_layers=args.disc_num_layers)
```
Comment (Lucaskabela) [inline src/train.py:891]: added a num layer argument here so we could get more representation :)

**Review by liaojh1998 — COMMENTED** [review]
I read through the code on the lengths for padding and I think they're okay, unless something is missing. Bill can do a third check on this too.

This looks good. Thanks for working on this ya'll!

File: src/network.py [hunk 3]
```diff
@@ -178,17 +178,12 @@ def __init__(self, d_in, hidden, bidirectional=False, num_layers=1, dropout=.2,
...
     def forward(self, out, out_len):
         _, (e_h, _) = self.rnn(out, out_len)
-        if self.num_dir == 2:
-            h = e_h.view(self.num_layers, self.num_dir, -1, self.hidden)
-            out = torch.cat((h[:, 0, :, :], h[:, 1, :, :]), dim=-1)
-        else:
-            out = e_h[-1]
-        return self.fc2(self.dropout(self.non_linear(self.reduce_h_W(out))))
+        # -1 gets topmost layer I think
+        return self.fc2(self.dropout(self.non_linear(e_h[-1])))
```
Comment (liaojh1998) [inline src/network.py]: The output is "h_n of shape (num_layers * num_directions, batch, hidden_size)" according to https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#torch.nn.LSTM. I don't think -1 can just do this though, but I should test it first.

Comment (liaojh1998) [reply src/network.py]: Nevermind, I thought RNN was an `LSTM`, but it turns out to be an `RNNEncoder`.

File: src/train.py [hunk 4]
```diff
@@ -469,17 +492,36 @@ def evaluate(model, valid_dataloader):
             x, _ = batch
             text, mel, text_len, mel_len = x
 
-            t_ae_loss, s_ae_loss = autoencoder_step(model, batch, args)
+            if args.use_discriminator:
```
Comment (billyang98) [inline src/train.py:496]: could probably just move everything to a single if else block

File: src/train.py [hunk 5]
```diff
@@ -268,67 +267,98 @@ def crossmodel_step(model, batch, args, use_dis_loss=False):
...
+        s_out = s_hid
+
+    # Construct targets
+    t_target = discriminator_target(t_out.shape[0], 'text').to(t_out)
+    s_target = discriminator_target(s_out.shape[0], 'speech').to(s_out)
+
+    # Pad both hidden states' sequence length with PAD_IDX to match dims
+    # Note: F.pad is nondeterministic in backward pass when used with CUDA,
+    #       but this should be ignored with pack_padded_sequence()?
+    t_seq_dim = t_out.shape[1]
```
Comment (billyang98) [inline src/train.py:309]: looks good! def would be good to sanity check

**Review by billyang98 — APPROVED** [review]
